arch/arm/src/common/arm_backtrace_unwind.c

```diff
@@ -0,0 +1,680 @@
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+#include <nuttx/arch.h>
+
+#include <arch/elf.h>
+
+#include "sched/sched.h"
+#include "arm_internal.h"
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+enum regs
+{
+#ifdef CONFIG_ARM_THUMB
+  FP = 7,
+#else
+  FP = 11,
+#endif /* CONFIG_ARM_THUMB */
+  SP = 13,
+  LR = 14,
+  PC = 15
+};
+
+/****************************************************************************
+ * Private Data Types
+ ****************************************************************************/
+
+struct unwind_frame_s
+{
+  unsigned long fp;
+  unsigned long sp;
+  unsigned long lr;
+  unsigned long pc;
+
+  /* address of the LR value on the stack */
+
+  unsigned long *lr_addr;
+
+  /* highest value of sp allowed */
+
+  unsigned long stack_top;
+};
+
+struct unwind_ctrl_s
+{
+  unsigned long        vrs[16];   /* virtual register set */
+  const unsigned long *insn;      /* pointer to the current instructions word */
+  unsigned long        stack_top; /* highest value of sp allowed */
+  unsigned long       *lr_addr;   /* address of LR value on the stack */
+  int                  entries;   /* number of entries left to interpret */
+  int                  byte;      /* current byte number in the instructions word */
+
+  /* 1 : check for stack overflow for each register pop.
+   * 0 : save overhead if there is plenty of stack remaining.
+   */
+
+  int check_each_pop;
+};
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/* Convert a prel31 symbol to an absolute address */
+
+#define prel31_to_addr(ptr)  \
+  ({  \
+  /* sign-extend to 32 bits */  \
+  long offset = (((long)*(ptr)) << 1) >> 1;  \
+  (unsigned long)(ptr) + offset;  \
+  })
+
+/****************************************************************************
+ * Name: search_index
+ *
+ * Description:
+ *  Binary search in the unwind index. The entries are
+ *  guaranteed to be sorted in ascending order by the linker.
+ *
+ *  start    = first entry
+ *  origin   = first entry with positive offset
+ *             (or stop if there is no such entry)
+ *  stop - 1 = last entry
+ *
+ ****************************************************************************/
+
+nosanitize_address
+static const struct __EIT_entry *
+search_index(unsigned long addr, const struct __EIT_entry *start,
+             const struct __EIT_entry *origin,
+             const struct __EIT_entry *stop)
+{
+  unsigned long addr_prel31;
+
+  /* only search in the section with the matching sign. This way the
```

Comment:
Change all first char in comment to upper case

Reply:
Done